airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_source.py

```diff
@@ -0,0 +1,165 @@
+
+
+class ConcurrentSource:
+    """
+    A Source that reads data from multiple AbstractStreams concurrently.
+    It does so by submitting partition generation, and partition read tasks to a thread pool.
+    The tasks asynchronously add their output to a shared queue.
+    The read is done when all partitions for all streams were generated and read.
+    """
+
+    DEFAULT_TIMEOUT_SECONDS = 900
+
+    @staticmethod
+    def create(
+        num_workers: int,
+        initial_number_of_partitions_to_generate: int,
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        message_repository: MessageRepository,
+        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
+    ) -> "ConcurrentSource":
+        threadpool = ThreadPoolManager(
+            concurrent.futures.ThreadPoolExecutor(max_workers=num_workers, thread_name_prefix="workerpool"), logger, num_workers
+        )
+        return ConcurrentSource(
+            threadpool, logger, slice_logger, message_repository, initial_number_of_partitions_to_generate, timeout_seconds
+        )
+
+    def __init__(
+        self,
+        threadpool: ThreadPoolManager,
+        logger: logging.Logger,
+        slice_logger: SliceLogger = DebugSliceLogger(),
+        message_repository: MessageRepository = InMemoryMessageRepository(),
+        initial_number_partitions_to_generate: int = 1,
+        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
+        raise_exception_on_missing_stream: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        """
+        :param threadpool: The threadpool to submit tasks to
+        :param message_repository: The repository to emit messages to
+        :param max_number_of_partition_generator_in_progress: The initial number of concurrent partition generation tasks. Limiting this number ensures will limit the latency of the first records emitted. While the latency is not critical, emitting the records early allows the platform and the destination to process them as early as possible.
+        :param timeout_seconds: The maximum number of seconds to wait for a record to be read from the queue. If no record is read within this time, the source will stop reading and return.
+        :param kwargs:
+        """
+        super().__init__(**kwargs)
+        self._threadpool = threadpool
+        self._logger = logger
+        self._slice_logger = slice_logger
+        self._message_repository = message_repository
+        self._initial_number_partitions_to_generate = initial_number_partitions_to_generate
+        self._timeout_seconds = timeout_seconds
+        self._raise_exception_on_missing_stream = raise_exception_on_missing_stream
+
+    def read(
+        self,
+        streams: List[AbstractStream],
+    ) -> Iterator[AirbyteMessage]:
+        self._logger.info("Starting syncing")
+        stream_instances_to_read_from = self._get_streams_to_read_from(streams)
```

Comment:
Given that availability checks can be time-consuming, is there any reason not to do this part concurrently? We can just put all the streams in the queue and let them be filtered out by the worker that's processing them.

That would make cost of the initial enqueuing of streams negligible and would allow us to simplify this method - we wouldn't need a separate `_submit_initial_partition_generators` step.